File: packages/google_mlkit_document_scanner/android/src/main/java/com/google_mlkit_document_scanner/DocumentScanner.java
package com.google_mlkit_document_scanner;

import android.app.Activity;
import android.content.Intent;
import android.content.IntentSender;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.mlkit.vision.documentscanner.GmsDocumentScanner;
import com.google.mlkit.vision.documentscanner.GmsDocumentScanning;
import com.google.mlkit.vision.documentscanner.GmsDocumentScannerOptions;
import com.google.mlkit.vision.documentscanner.GmsDocumentScanningResult;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

public class DocumentScanner implements MethodChannel.MethodCallHandler, PluginRegistry.ActivityResultListener {
    private static final String START = "vision#startDocumentScanner";
    private static final String CLOSE = "vision#closeDocumentScanner";
    private static final String TAG = "DocumentScanner";
    private final Map<String, GmsDocumentScanner> instances = new HashMap<>();
    private final ActivityPluginBinding binding;
    private MethodChannel.Result pendingResult = null;
    final private int START_DOCUMENT_ACTIVITY = 0x362738;

    public DocumentScanner(ActivityPluginBinding binding) {
        this.binding = binding;
        binding.addActivityResultListener(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        switch (method) {
            case START:
                handleScanner(call, result);
                break;
            case CLOSE:
                closeScanner(call);
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void handleScanner(MethodCall call, final MethodChannel.Result result) {
        String id = call.argument("id");
        GmsDocumentScanner scanner = instances.get(id);
        pendingResult = result;

        // Create a new scanner instance if it doesn't exist
        if (scanner == null) {
            Map<String, Object> options = call.argument("options");
            if (options == null) {
                result.error(TAG, "Invalid options", null);
                return;
            }
            GmsDocumentScannerOptions scannerOptions = parseOptions(options);
            scanner = GmsDocumentScanning.getClient(scannerOptions);
            instances.put(id, scanner);
        }

        Activity activity = binding.getActivity();
        scanner.getStartScanIntent(activity).addOnSuccessListener(new OnSuccessListener<IntentSender>() {
            @Override
            public void onSuccess(IntentSender intentSender) {
                try {
                    activity.startIntentSenderForResult(intentSender, START_DOCUMENT_ACTIVITY, null, 0, 0, 0);
                } catch (IntentSender.SendIntentException e) {
                    result.error(TAG, "Failed to start document scanner", null);
                }
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                result.error(TAG, "Failed to start document scanner", null);
            }
        });
    }

    // parse scanner options
    private GmsDocumentScannerOptions parseOptions(Map<String, Object> options) {
        boolean isGalleryImportAllowed = (boolean) options.get("isGalleryImport");
        int pageLimit = (int) options.get("pageLimit");
        int format;
        switch ((String) Objects.requireNonNull(options.get("format"))) {
            case "pdf":
                format = GmsDocumentScannerOptions.RESULT_FORMAT_PDF;
                break;
            case "jpeg":
                format = GmsDocumentScannerOptions.RESULT_FORMAT_JPEG;
                break;
            default:
                throw new IllegalArgumentException("Not a format:" + options.get("format"));
        }
        int mode;
        switch ((String) options.get("mode")) {
            case "base":
                mode = GmsDocumentScannerOptions.SCANNER_MODE_BASE;
                break;
            case "filter":
                mode = GmsDocumentScannerOptions.SCANNER_MODE_BASE_WITH_FILTER;
                break;
            case "full":
                mode = GmsDocumentScannerOptions.SCANNER_MODE_FULL;
                break;
            default:
                throw new IllegalArgumentException("Not a mode:" + options.get("mode"));
        }
        GmsDocumentScannerOptions.Builder builder = new GmsDocumentScannerOptions.Builder().setGalleryImportAllowed(isGalleryImportAllowed).setPageLimit(pageLimit).setResultFormats(format).setScannerMode(mode);
        return builder.build();
    }

    private void closeScanner(MethodCall call) {
        String id = call.argument("id");
        GmsDocumentScanner scanner = instances.get(id);
        if (scanner == null) return;
        instances.remove(id);
    }

    @Override
    public boolean onActivityResult(int requestCode, int resultCode, @Nullable Intent intent) {
        if (requestCode == START_DOCUMENT_ACTIVITY) {
            if (resultCode == Activity.RESULT_OK) {
                GmsDocumentScanningResult result = GmsDocumentScanningResult.fromActivityResultIntent(intent);
                if (result != null) {
                    handleScanningResult(result);
                }
            } else if (resultCode == Activity.RESULT_CANCELED) {
                pendingResult.error(TAG, "Operation cancelled", null);
            } else {
                pendingResult.error(TAG, "Unknown Error", null);
            }
            return true;
        }
        return false;
    }

    private void handleScanningResult(GmsDocumentScanningResult result) {
        Map<String, Object> resultMap = new HashMap<>();

        // Check if the result has a pdf
        GmsDocumentScanningResult.Pdf pdf = result.getPdf();
        if (pdf != null) {
            Map<String, Object> pdfMap = new HashMap<>();
            pdfMap.put("pageCount", pdf.getPageCount());
            pdfMap.put("uri", pdf.getUri().getPath());
            resultMap.put("pdf", pdfMap);
        } else {
            resultMap.put("pdf", null);
        }

        // Check if the result has a list of pages
        List<GmsDocumentScanningResult.Page> pages = result.getPages();
        if (pages != null && !pages.isEmpty()) {
            List<String> imageUris = new ArrayList<>();
            for (GmsDocumentScanningResult.Page page : pages) {
                imageUris.add(page.getImageUri().getPath());
            }
            resultMap.put("images", imageUris);
        } else {
            resultMap.put("images", null);
        }

        pendingResult.success(resultMap);
    }
}

File: packages/google_mlkit_document_scanner/android/src/main/kotlin/com/google_mlkit_document_scanner/DocumentScanner.kt
package com.google_mlkit_document_scanner

import android.app.Activity
import android.content.Intent
import android.content.IntentSender
import com.google.mlkit.vision.documentscanner.GmsDocumentScanner
import com.google.mlkit.vision.documentscanner.GmsDocumentScannerOptions
import com.google.mlkit.vision.documentscanner.GmsDocumentScanning
import com.google.mlkit.vision.documentscanner.GmsDocumentScanningResult
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class DocumentScanner(
    private val binding: ActivityPluginBinding,
) : MethodChannel.MethodCallHandler,
    PluginRegistry.ActivityResultListener {
    private val instances = HashMap<String, GmsDocumentScanner>()
    private var pendingResult: MethodChannel.Result? = null

    companion object {
        private const val START = "vision#startDocumentScanner"
        private const val CLOSE = "vision#closeDocumentScanner"
        private const val TAG = "DocumentScanner"
        private const val START_DOCUMENT_ACTIVITY = 0x362738
    }

    init {
        binding.addActivityResultListener(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            START -> {
                handleScanner(call, result)
            }

            CLOSE -> {
                closeScanner(call)
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun handleScanner(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val id = call.argument<String>("id")
        var scanner = instances[id]
        pendingResult = result

        if (scanner == null) {
            val options =
                call.argument<Map<String, Any>>("options") ?: run {
                    consumePendingResult()?.error(TAG, "Invalid options", null)
                    return
                }
            val scannerOptions = parseOptions(options)
            scanner = GmsDocumentScanning.getClient(scannerOptions)
            instances[id!!] = scanner
        }

        val activity = binding.activity
        scanner
            .getStartScanIntent(activity)
            .addOnSuccessListener { intentSender ->
                try {
                    activity.startIntentSenderForResult(intentSender, START_DOCUMENT_ACTIVITY, null, 0, 0, 0)
                } catch (e: IntentSender.SendIntentException) {
                    consumePendingResult()?.error(TAG, "Failed to start document scanner", null)
                }
            }.addOnFailureListener {
                consumePendingResult()?.error(TAG, "Failed to start document scanner", null)
            }
    }

    private fun parseOptions(options: Map<String, Any>): GmsDocumentScannerOptions {
        val isGalleryImportAllowed = options["isGalleryImport"] as Boolean
        val pageLimit = options["pageLimit"] as Int
        val formatStrings = options["formats"] as List<*>

        val formatConstants =
            formatStrings.map { format ->
                when (format) {
                    "pdf" -> GmsDocumentScannerOptions.RESULT_FORMAT_PDF
                    "jpeg" -> GmsDocumentScannerOptions.RESULT_FORMAT_JPEG
                    else -> throw IllegalArgumentException("Not a format: $format")
                }
            }

        val mode =
            when (options["mode"] as String) {
                "base" -> GmsDocumentScannerOptions.SCANNER_MODE_BASE
                "filter" -> GmsDocumentScannerOptions.SCANNER_MODE_BASE_WITH_FILTER
                "full" -> GmsDocumentScannerOptions.SCANNER_MODE_FULL
                else -> throw IllegalArgumentException("Not a mode: ${options["mode"]}")
            }

        val builder =
            GmsDocumentScannerOptions
                .Builder()
                .setGalleryImportAllowed(isGalleryImportAllowed)
                .setPageLimit(pageLimit)
                .setScannerMode(mode)

        if (formatConstants.isNotEmpty()) {
            if (formatConstants.size > 1) {
                builder.setResultFormats(formatConstants[0], formatConstants[1])
            } else {
                builder.setResultFormats(formatConstants[0])
            }
        }

        return builder.build()
    }

    private fun closeScanner(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        instances.remove(id)
    }

    override fun onActivityResult(
        requestCode: Int,
        resultCode: Int,
        intent: Intent?,
    ): Boolean {
        if (requestCode == START_DOCUMENT_ACTIVITY) {
            when (resultCode) {
                Activity.RESULT_OK -> {
                    val result = GmsDocumentScanningResult.fromActivityResultIntent(intent)
                    result?.let { handleScanningResult(it) }
                }

                Activity.RESULT_CANCELED -> {
                    consumePendingResult()?.error(TAG, "Operation cancelled", null)
                }

                else -> {
                    consumePendingResult()?.error(TAG, "Unknown Error", null)
                }
            }
            return true
        }
        return false
    }

    private fun handleScanningResult(result: GmsDocumentScanningResult) {
        val reply = consumePendingResult() ?: return

        val resultMap = HashMap<String, Any?>()

        val pdf = result.pdf
        if (pdf != null) {
            resultMap["pdf"] =
                hashMapOf(
                    "pageCount" to pdf.pageCount,
                    "uri" to pdf.uri.path,
                )
        } else {
            resultMap["pdf"] = null
        }

        val pages = result.pages
        if (!pages.isNullOrEmpty()) {
            resultMap["images"] = pages.map { it.imageUri.path }
        } else {
            resultMap["images"] = null
        }

        reply.success(resultMap)
    }

    /**
     * Atomically returns the current [pendingResult] and clears the field, so a subsequent
     * duplicate callback (e.g. a redelivered activity result on some OEM builds) cannot submit
     * a second reply on the same [MethodChannel.Result] and trigger
     * `IllegalStateException: Reply already submitted`.
     */
    private fun consumePendingResult(): MethodChannel.Result? {
        val reply = pendingResult
        pendingResult = null
        return reply
    }
}
